src/writ/transform/svdframe.py

```diff
@@ -0,0 +1,138 @@
+    of the iterates in-place. The positions of the coordinates and forces in iterate
+    can be specified at initialization.
+    """
+
+    def __init__(
+        self,
+        source: Iterable[Sequence[A]],
+        coords_idx: int = 0,
+        forces_idx: int = 1,
+        stride_for_svd: int = 1,
+    ) -> None:
+        """Initialize.
+
+        Arguments:
+        ---------
+        source:
+            Iterable of objects to iterate over
+        coords_idx: int
+            Index of the coordinates array in the tuples we iterate over,
+            by default 0.
+        forces_idx: int
+            Index of the forces array in the tuples we iterate over,
+            by default 1.
+        stride_for_svd: int
+            Stride to use when choosing atoms for SVD fit,
+            by default 1.
+            Note that if the stride would result in using less than 3 particles,
+            we will fallback to a stride of 1.
+
+        """
+        self.source = source
+        self.coords_idx = coords_idx
+        self.forces_idx = forces_idx
+        self.stride_for_svd = stride_for_svd
+
+    def _rotate(
+        self, coords_batch: ArrayMxNx3, forces_batch: ArrayMxNx3
+    ) -> Tuple[ArrayMxNx3, ArrayMxNx3]:
+        rotated_coords = []
+        rotated_forces = []
+        for coords, forces in zip(coords_batch, forces_batch):
+            c_rot, f_rot = self._rotate_single_frame(coords=coords, forces=forces)
+            rotated_coords += [c_rot]
+            rotated_forces += [f_rot]
+        return (
+            np.asarray(rotated_coords),
+            np.asarray(rotated_forces),
+        )
+
+    def _rotate_single_frame(
+        self, coords: ArrayNx3, forces: ArrayNx3
+    ) -> Tuple[ArrayNx3, ArrayNx3]:
+        # center the molecule around the mean
+        coords = coords - coords.mean(0)
+        # need at least 3 atoms to compute SVD frame
+        if coords.shape[0] // self.stride_for_svd < 3:
+            coords_subset = coords
+        else:
+            coords_subset = coords[:: self.stride_for_svd]
+        # (again) make sure we have enough atoms to calculate svd frame
+        if coords_subset.shape[0] > 2:
+            U, S, Vh = np.linalg.svd(
+                coords_subset, full_matrices=False, compute_uv=True
+            )
+            V = Vh.T
+            # NOTE: np.matmul is slower than mat1 @ mat2...
+            rotated_coords = coords @ V  # np.matmul(coords, V)
+            rotated_forces = forces @ V  # np.matmul(forces, V)
+        else:
+            rotated_coords = coords
+            rotated_forces = forces
+        return rotated_coords, rotated_forces
+
+    def _check_coords_forces_shapes_match(
+        self, coords_batch: np.ndarray, forces_batch: np.ndarray
+    ) -> TypeGuard[ArrayMxNx3]:
+        coords_shape = coords_batch.shape
+        forces_shape = forces_batch.shape
+        assert len(coords_shape) == 3
```

Comment:
change these asserts to ValueErrors so that you can give a error message

Reply:
It removes the ability to turn off the asserts for performance., but it will help usage later. If you want you could add a config flag to turn checks on or off.